storage/src/tree_state.rs

```diff
@@ -0,0 +1,396 @@
+		H256::from("7e4320434849ecb357f1afaaba21a54400ef2d11cff83b937d87fdafa49f8199"),
+		H256::from("020adc98d96cfbbcca15fc3aa03760ed286686c35b5d92c7cb64a999b394a854"),
+		H256::from("3a26b29fe1acfdd6c6a151bcc3dbcb95a10ebe2f0553f80779569b67b7244e77"),
+		H256::from("ec2d0986e6a0ddf43897b2d4f23bb034f538ffe00827f310dc4963f3267f0bfb"),
+		H256::from("d48073f8819f81f0358e3fc35a047cc74082ae1cb7ee22fb609c01649342d0e6"),
+		H256::from("ad8037601793f172441ecb00dc138d9fc5957125ecc382ec65e36f817dc799fb"),
+		H256::from("ca500a5441f36f4df673d6b8ed075d36dae2c7e6481428c70a5a76b7a9bebce8"),
+		H256::from("422b6ddd473231dc4d56fe913444ccd56f7c61f747ba57ca946d5fef72d840a0"),
+		H256::from("ab41f4ecb7d7089615800e19fcc53b8379ed05ee35c82567095583fd90ff3035"),
+		H256::from("bbf7618248354ceb1bc1fc9dbc42c426a4e2c1e0d443c5683a9256c62ecdc26f"),
+		H256::from("e50ae71479fc8ec569192a13072e011afc249f471af09500ea39f75d0af856bf"),
+		H256::from("e74c0b9220147db2d50a3b58d413775d16c984690be7d90f0bc43d99dba1b689"),
+		H256::from("29324a0a48d11657a51ba08b004879bfcfc66a1acb7ce36dfe478d2655484b48"),
+		H256::from("88952e3d0ac06cb16b665201122249659a22325e01c870f49e29da6b1757e082"),
+		H256::from("cdf879f2435b95af042a3bf7b850f7819246c805285803d67ffbf4f295bed004"),
+		H256::from("e005e324200b4f428c62bc3331e695c373607cd0faa9790341fa3ba1ed228bc5"),
+		H256::from("354447727aa9a53dd8345b6b6c693443e56ef4aeba13c410179fc8589e7733d5"),
+		H256::from("da52dda91f2829c15c0e58d29a95360b86ab30cf0cac8101832a29f38c3185f1"),
+		H256::from("c7da7814e228e1144411d78b536092fe920bcdfcc36cf19d1259047b267d58b5"),
+		H256::from("aba1f68b6c2b4db6cc06a7340e12313c4b4a4ea6deb17deb3e1e66cd8eacf32b"),
+		H256::from("c160ae4f64ab764d864a52ad5e33126c4b5ce105a47deedd75bc70199a5247ef"),
+		H256::from("eadf23fc99d514dd8ea204d223e98da988831f9b5d1940274ca520b7fb173d8a"),
+		H256::from("5b8e14facac8a7c7a3bfee8bae71f2f7793d3ad5fe3383f93ab6061f2a11bb02")
+	].to_vec();
+}
+
+pub trait Dim {
+	const HEIGHT: usize;
+}
+
+pub struct H32;
+
+impl Dim for H32 {
+	const HEIGHT: usize = 32;
+}
+
+pub struct TreeState<D: Dim> {
+	_phantom: ::std::marker::PhantomData<D>,
+	left: Option<H256>,
+	right: Option<H256>,
+	parents: Vec<Option<H256>>,
+}
+
+impl<D: Dim> TreeState<D> {
+	pub fn new() -> Self {
+		TreeState {
+			_phantom: ::std::marker::PhantomData,
+			left: None,
+			right: None,
+			parents: vec![None; D::HEIGHT - 1],
+		}
+	}
+
+	pub fn append(&mut self, hash: H256) -> Result<(), &'static str> {
+		if self.left.is_none() {
+			self.left = Some(hash);
+		} else if self.right.is_none() {
+			self.right = Some(hash);
+		} else {
+			let former_left = ::std::mem::replace(&mut self.left, Some(hash))
+				.expect("none variant is handled in the branch above; qed");
+			let former_right = self.right.take()
+				.expect("none variant is handled in the branch above; qed");
+
+			let mut combined = sha256_compress(&*former_left, &*former_right);
+			for i in 0..D::HEIGHT-1 {
```

Comment:
Are you sure it should be `D::HEIGHT-1` instead of `D::HEIGHT`?

Reply:
Also in the constructor: `parents: vec![None; D::HEIGHT - 1],`

Reply:
Yeah quite sure, one level does not need to be stored o calculated in any way